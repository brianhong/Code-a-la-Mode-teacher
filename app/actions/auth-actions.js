// @flow
import axios from 'axios';
import gitAuth from '../utils/github.settings';

export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';
export const LOGIN_FAILURE = 'LOGIN_FAILURE';
export const LOGOUT = 'LOGOUT';

type authInfoType = {
  client_id: string,
  client_secret: string,
  authCode: string
};

const postCredentials = (authInfo: authInfoType) => {
  const url = 'https://github.com/login/oauth/access_token';
  const config = {
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
      'Cache-Control': 'no-cache'
    }
  };
  return axios.post(url, authInfo, config);
};

export const teacherLogin = (authCode: string) => (dispatch: *) => {
  const authInfo = {
    client_id: gitAuth.client_id,
    client_secret: gitAuth.secret,
    authCode
  };

  return postCredentials(authInfo)
    .then(response => dispatch({ type: LOGIN_SUCCESS, token: response.data.access_token }))
    .catch(error => {
      console.error(error);
      dispatch({ type: LOGIN_FAILURE, payload: error.response.data });
    });
};

export const logout = () => ({ type: LOGOUT });

// export const loginUser = (code) => {
//   return (dispatch, getState) => {

//     const url = 'https://github.com/login/oauth/access_token';
//     const method = 'POST';
//     const data = {
//       client_id: options.client_id,
//       client_secret: options.secret,
//       code: code
//     };

//     dispatch({type: LOGIN.REQUEST});

//     return apiRequest(url, method, data)
//       .then((response) => {
//         return dispatch({type: LOGIN.SUCCESS, payload: response.data});
//       })
//       .then(fetchUsername())
//       .catch((error) => {
//         console.error(error);
//         dispatch({type: LOGIN.FAILURE, payload: error.response.data});
//       });

//   };
// }
