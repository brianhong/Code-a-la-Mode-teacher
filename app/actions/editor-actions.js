// @flow
import { ipcRenderer } from 'electron';

export const EDITOR_CHANGE = 'EDITOR_CHANGE';
export const editorChange = (contents: string) => ({ type: EDITOR_CHANGE, contents });

type actionType = {
  type: string
};

export function changeEditor(contents: string) {
  ipcRenderer.send('editor-change', contents);
  console.log('changeEditor contents: ', contents);
  return (dispatch: (action: actionType) => void) => {
    dispatch(editorChange(contents));
  };
}
