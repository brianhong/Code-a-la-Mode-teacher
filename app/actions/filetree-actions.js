import Username from 'username';
import { ipcRenderer } from 'electron';
import { readFile } from '../utils/FileSystemUtils';

export const GOT_USERNAME = 'GOT_USERNAME';
export const OPEN_FILE = 'OPEN_FILE';
export const XMIT_FILE = 'XMIT_FILE';
export const FILETREE_CHANGE = 'FILETREE_CHANGE';

type actionType = {
  type: string
};

export const gotUsername = username => ({ type: GOT_USERNAME, username });
export const openFile = newEditorState => ({ type: OPEN_FILE, newEditorState });
export const xmitFile = contents => ({ type: XMIT_FILE, contents });
export const filetreeChange = (filetree: Object[]) => ({ type: FILETREE_CHANGE, filetree });

export function getUsername() {
  return (dispatch: (action: actionType) => void) => {
    Username().then(name => dispatch(gotUsername(name)))
      .catch(console.error);
  };
}

export function loadFile(selectedFile, currentOpenFiles, currentEditorValues, selectedFileIndex) {
  return (dispatch: (action: actionType) => void) => {
    const loadedFilePath = selectedFile.filePath;

    // Do not try to load a file already inside the Editor
    if (currentOpenFiles.includes(loadedFilePath)) {
      // ADD CODE TO DISPATCH ACTION TO OPEN THE UPDATED VALUE OF THE FILE
      return;
    }

    console.log(`SELECTED FILE INDEX BEFORE: ${selectedFileIndex} AFTER ${selectedFileIndex + 1}`);
    // Opening a new file, load it into the Editor
    readFile(selectedFile.filePath)
      .then(contents => {
        const allOpenFiles = currentOpenFiles.concat(loadedFilePath);
        const text = contents.toString();
        const newEditorValues = currentEditorValues.concat([text]);
        const newEditorState = {
          contents: newEditorValues,
          currentOpenFiles: allOpenFiles,
          selectedFileIndex: selectedFileIndex + 1
        };
        return dispatch(openFile(newEditorState));
      })
      .catch(console.error);
  };
}

export function reqAndXmitFile(filePath) {
  readFile(filePath)
    .then(contents => {
      const text = contents.toString();
      return ipcRenderer.send('xmit-file', text);
    })
    .catch(console.error);
}

export function setFiletree(tree: Object) {
  ipcRenderer.send('xmit-tree', tree);
  return filetreeChange(tree);
}
