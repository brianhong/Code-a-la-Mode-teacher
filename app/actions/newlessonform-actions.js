// @flow
export const LESSON_NAME_INPUT = 'LESSON_NAME_INPUT';

export const lessonNameChange = (event) => (dispatch: *) => {
  dispatch({ type: LESSON_NAME_INPUT, newLessonName: event.target.value });
}

