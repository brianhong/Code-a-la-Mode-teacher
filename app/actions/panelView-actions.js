// @flow
export const VIEW_TICKETS = 'VIEW_TICKETS';
export const VIEW_HISTORY = 'VIEW_HISTORY';

export function viewTickets() {
  return { type: VIEW_TICKETS };
}

export function viewHistory() {
  return { type: VIEW_HISTORY };
}
