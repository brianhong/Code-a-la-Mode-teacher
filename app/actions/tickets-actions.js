// @flow
export const ADD_TICKET = 'ADD_TICKET';
export const SELECT_TICKET = 'SELECT_TICKET';
export const REMOVE_TICKET = 'REMOVE_TICKET';

// codestate subject to change- still unsure how we are storing
// current state of code
// student-side will need some form of create-ticket
// discussion: what kinds of props should a ticket have?
export type ticketType = {
  question: string,
  takeSnapshot?: boolean
};

// may add a takeSnapshot? property to ticket later
export function add(question: string) {
  return {
    type: ADD_TICKET,
    ticket: { question }
  };
}

export function select(ticket: ticketType) {
  return {
    type: SELECT_TICKET,
    ticket
  };
}

export function remove(ticket: ticketType) {
  return {
    type: REMOVE_TICKET,
    ticket
  };
}
