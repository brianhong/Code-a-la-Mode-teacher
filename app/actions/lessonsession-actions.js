import { remote } from 'electron';
import axios from 'axios';
import PromiseB from 'bluebird';

import { lessonInfoType } from '../reducers/lessonSession-reducer';
import { makeDirectory, writeFile } from '../utils/FileSystemUtils';
import { getLastFromPath } from '../utils/file-functions';

import { clearEditorState } from './editor-actions';

const git = require('simple-git');

const GITHUB_API_ROOT = 'https://api.github.com';

export const LOAD_USER_REPOS = 'LOAD_USER_REPOS';
export const LOAD_LESSON = 'LOAD_LESSON';
export const CHECKOUT_NEXT_BRANCH = 'CHECKOUT_NEXT_BRANCH';
export const CHECKOUT_PREVIOUS_BRANCH = 'CHECKOUT_PREVIOUS_BRANCH';
export const CANNOT_CHECKOUT = 'CANNOT_CHECKOUT';
export const ADD_BRANCH = 'ADD_BRANCH';
export const CREATED_NEW_LESSON = 'CREATED_NEW_LESSON';
export const ADD_HEAD_HASH = 'ADD_HEAD_HASH';

export const loadUserRepos = (userRepositories: []) => (dispatch: *) => {
  dispatch({ type: LOAD_USER_REPOS, userRepositories });
};

export const loadAfterCreating = (lessonFilePath: string) => (dispatch: *) => {
  const lessonGit = git(lessonFilePath);
  lessonGit
    .branch((err, branchSummary) => {
      const branchNames = Object.keys(branchSummary.branches).map((branchName) => branchName);

      dispatch({
        type: CREATED_NEW_LESSON,
        lessonInfo: {
          branches: branchSummary.branches,
          branchIndex: 0,
          repositoryPath: lessonFilePath,
          branchNames: branchNames,
          currentBranch: branchSummary.current,
          headHashes: {}
        } });
    });
};


export const loadAfterCloning = (lessonFilePath: string) => (dispatch: *) => {
  const lessonGit = git(lessonFilePath);

  lessonGit
    .branch((err, branchSummary) => {
      const checkoutPromises = [];
      const branchNames = Object.keys(branchSummary.branches).map((branchName) => {
        const localBranchName = getLastFromPath(branchName);

        // Handling remote branches for local working copy
        if (localBranchName !== 'master') {
          checkoutPromises.push(
            new Promise((resolve, reject) => {
              lessonGit
                .checkoutBranch(localBranchName, branchName, (checkoutError, success) => {
                  // Not really using success, just care about checkoutError
                  return checkoutError ? console.error(checkoutError) : localBranchName;
                })
                .exec(resolve());
            }
          ));
        }
        return localBranchName;
      });

      return PromiseB.all(checkoutPromises)
        .then(() => {
          lessonGit
            .checkout('master')
            .exec(() => {
              dispatch({
                type: LOAD_LESSON,
                lessonInfo: {
                  branches: branchSummary.branches,
                  branchIndex: 0,
                  repositoryPath: lessonFilePath,
                  branchNames: branchNames,
                  currentBranch: branchSummary.current,
                  headHashes: {}
                }
              });
            });
        });
    });
};

export const checkoutNextBranch = (lessonInfo: lessonInfoType,
                                   currentOpenFiles: Array<string>,
                                   currentEditorValues: Array<string>) => (dispatch: *) => {
  const currentIndex = lessonInfo.branchIndex;
  if (currentIndex === lessonInfo.branches.length) {
    dispatch({
      type: CANNOT_CHECKOUT
    });
    return;
  }

  saveLesson(currentOpenFiles, currentEditorValues)
    .then(() => {
      const nextBranchName = lessonInfo.branchNames[currentIndex + 1];

      git(lessonInfo.repositoryPath)
        .add('./*')
        .commit(`Temporary commit for branch ${lessonInfo.branchNames[currentIndex]}`)
        .checkout(nextBranchName, (err) => {
          if (err) {
            dispatch({ type: CANNOT_CHECKOUT });
          } else {
            // First time seeing a branch, store the commit hash of the HEAD
            if (!Object.prototype.hasOwnProperty.call(lessonInfo.headHashes, nextBranchName)) {
              git(lessonInfo.repositoryPath)
                .revparse(['master'], (error, headHash) => {
                  const newHeadHash = { [nextBranchName]: headHash };
                  dispatch({ type: ADD_HEAD_HASH, newHeadHash });
                });
            }

            dispatch(clearEditorState());

            dispatch({
              type: CHECKOUT_NEXT_BRANCH,
              currentBranch: nextBranchName,
              newBranchIndex: currentIndex + 1
            });
          }
        });
    })
    .catch(error => console.error(error));
};

export const checkoutPreviousBranch = (lessonInfo: lessonInfoType,
                                        currentOpenFiles: Array<string>,
                                        currentEditorValues: Array<string>) => (dispatch: *) => {
  const currentIndex = lessonInfo.branchIndex;
  if (currentIndex === 0) {
    dispatch({
      type: CANNOT_CHECKOUT
    });
    return;
  }

  saveLesson(currentOpenFiles, currentEditorValues)
    .then(() => {
      const previousBranchName = lessonInfo.branchNames[currentIndex - 1];

      git(lessonInfo.repositoryPath)
        .add('./*')
        .commit(`Temporary commit for branch ${lessonInfo.branchNames[currentIndex]}`)
        .checkout(previousBranchName, (err) => {
          if (err) {
            dispatch({ type: CANNOT_CHECKOUT });
          } else {
            // First time seeing a branch, store the commit hash of the HEAD
            if (!Object.prototype.hasOwnProperty.call(lessonInfo.headHashes, previousBranchName)) {
              git(lessonInfo.repositoryPath)
                .revparse(['master'], (error, headHash) => {
                  const newHeadHash = { [previousBranchName]: headHash };
                  dispatch({ type: ADD_HEAD_HASH, newHeadHash });
                });
            }

            dispatch(clearEditorState());

            dispatch({
              type: CHECKOUT_PREVIOUS_BRANCH,
              currentBranch: previousBranchName,
              newBranchIndex: currentIndex - 1
            });
          }
        });
    })
    .catch(error => console.error(error));
};

export const createNewLesson = (event: Object,
                                newLessonName: string,
                                userToken: string,
                                history) => (dispatch: *) => {
  event.preventDefault();
  remote.dialog.showSaveDialog({ defaultPath: newLessonName }, (newLessonFilePath) => {
    makeDirectory(newLessonFilePath)
      .then((err) => {
        if (err) {
          console.error(`Failed to create direcotry: ${err}`);
          return;
        }
        git(newLessonFilePath)
          .init()
          .commit('Initial commit');
      })
      .then(() => {
        const config = {
          headers: {
            Authorization: `token ${userToken}`,
          }
        };

        const data = {
          name: newLessonName,
          gitignore_template: 'Node'
        };

        axios.post(`${GITHUB_API_ROOT}/user/repos`, data, config)
          .then((response) => {
            console.log(`Repository successfully created with response status ${response.status}`);

            git(newLessonFilePath)
              .addRemote('origin', response.data.html_url)
              .pull('origin', 'master')
              .push('origin', 'master');

            dispatch(loadAfterCreating(newLessonFilePath));
            history.push('/editor');
          })
          .catch(error => {
            console.error(error);
          });
      })
      .catch(error => console.error(error));
  });
};

export const addBranch = (branch, branchName) => ({
  type: ADD_BRANCH,
  branch,
  branchName
});

export const saveLesson = (currentOpenFiles: Array<string>, currentEditorValues: Array<string>) => {
  const writePromises = currentOpenFiles.map((filePath, index) => {
    const newFileContent = currentEditorValues[index];
    return writeFile(filePath, newFileContent);
  });
  return Promise.all(writePromises);
};
