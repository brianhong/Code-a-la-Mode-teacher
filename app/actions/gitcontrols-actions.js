import axios from 'axios';
import { loadUserRepos } from './lessonsession-actions';

const git = require('simple-git')();

export const RECEIVED_USER_REPOS = 'RECEIVED_USER_REPOS';

const GITHUB_API_ROOT = 'https://api.github.com';

export const getUserRepositories = (userToken: string) => (dispatch: *) => {
  const config = {
    headers: {
      Authorization: `token ${userToken}`,
    },
    params: {
      affiliation: 'owner, collaborator',
      per_page: 300,
      sort: 'full_name'
    }
  };

  axios.get(`${GITHUB_API_ROOT}/user/repos`, config)
    .then(userRepos => {
      git
        .exec(() => {
          dispatch(loadUserRepos(userRepos.data));
        });
    })
    .catch(error => {
      console.error(error);
    });
};
