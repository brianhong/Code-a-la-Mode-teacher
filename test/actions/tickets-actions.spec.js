import * as actions from '../../app/actions/tickets-actions';

const testTicket = {
  id: 1,
  question: 'Is this test passing?'
};

// snapshot test workflow:
// first time a test with .toMatchSnapshot() is run, a literal copy
// of that function/test's output is saved as a snapshot.
// the next time the test is run, a diff is run against the snapshot
// if they do not match, an error is thrown and you are prompted to
// either update the snapshot, or change your code to match the old snapshot

// take a look at './__snapshots__/tickets-actions.spec.js.snap' to see
// the snapshot these created actions
describe('action creators', () => {
  it('should create an addAction', () => {
    expect(actions.add(testTicket)).toMatchSnapshot();
  });

  it('should create a selectAction', () => {
    expect(actions.select(testTicket)).toMatchSnapshot();
  });

  it('should create a removeAction', () => {
    expect(actions.remove(testTicket)).toMatchSnapshot();
  });
});
