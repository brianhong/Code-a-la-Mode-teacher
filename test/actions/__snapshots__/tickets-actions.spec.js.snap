// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action creators should create a removeAction 1`] = `
Object {
  "id": 1,
  "type": "REMOVE_TICKET",
}
`;

exports[`action creators should create a selectAction 1`] = `
Object {
  "id": 1,
  "type": "SELECT_TICKET",
}
`;

exports[`action creators should create an addAction 1`] = `
Object {
  "id": 1,
  "ticket": Object {
    "id": 1,
    "question": "Is this test passing?",
  },
  "type": "ADD_TICKET",
}
`;
